Synchronize snapshot and staging sstable deletion using sem

bhalevy opened the pull request:
Fixes #5340
(instead of #5329 that was retracted)

Hold the sstable_deletion_sem table::move_sstables_from_subdirs to serialize access to the staging directory.  It now synchronizes snapshot, compaction deletion of sstables, and view_update_generator moving of sstables from staging.

Tests:
* unit (dev) [expect  test_user_function_timestamp_return that fails for me locally, but also on master]
* snapshot_test.py (dev)

denesb commented on database.hh:
```diff
@@ -533,7 +533,7 @@ private:
     utils::phased_barrier _pending_streams_phaser;
```
This commit seems to do more than what is in the description, it now also syncs the touched directories.

bhalevy replied:
Why? `sstable::move_to_new_dir_in_thread` used to do it unconditionally before this change.

denesb commented on db/view/view_update_generator.cc:
```diff
@@ -39,10 +39,22 @@ future<> view_update_generator::start() {
+                    _sstables_to_move.emplace(t, entry);
+                } else {
+                    it->second.sstables.push_back(entry.sst);
+                }
```
Wouldn't `_sstables_to_move[t].push_back(entry.sst`) achieve the same?

bhalevy replied:
Yeah, with a default constructor for `sstables_to_move`.
Good idea...

bhalevy replied:
Hmm, we also need to set `_sstables_to_move[t].t = entry.t` at least once
so maybe it's more elegant but less efficient.
I want to hold on to a lw_shared_ptr<table> rather than just a `table*`.
If it would be the latter case, we could just use the `table*` key.

bhalevy replied:
done

denesb commented on db/view/view_update_generator.cc:
```diff
@@ -39,10 +39,22 @@ future<> view_update_generator::start() {
                 }
-                entry.t->move_sstable_from_staging_in_thread(entry.sst);
-                _registration_sem.signal();
+                auto t = entry.t.get();
```
This change is also not explained.

bhalevy replied:
right about `_registration_sem.signal()`
It was dropped inadvertently.

bhalevy replied:
fixed

bhalevy commented:
In v2:
- Resurrected `_registration_sem.signal()`
- Added a couple comments.

Test: unit(dev), unit/view_build_test(debug)

denesb commented on db/view/view_update_generator.cc:
```diff
@@ -40,21 +40,16 @@ future<> view_update_generator::start() {
             for (auto it = _sstables_to_move.begin(); it != _sstables_to_move.end(); ) {
-                auto& entry = it->second;
+                auto& t = it->first;
+                auto& sstables = it->second;
```
Could also been written as:
```
auto& [t, sstables] = *it;
```

denesb replied:
(no need to change)

denesb commented on db/view/view_update_generator.cc:
```diff
@@ -39,10 +39,25 @@ future<> view_update_generator::start() {
+                } else {
+                    it->second.sstables.push_back(entry.sst);
+                }
                 _registration_sem.signal();
```
If the `push_back()` above throws this will be leaked.

bhalevy replied:
That was true also before the change, for move_sstable_from_staging_in_thread (and to read_rows_flat and consume_in_thread), no?

denesb replied:
Yes, there is preexisting potentially throwing code above. We still probably should fix this although not in this patch. My gut feeling is that not releasing this semaphore will convert some exception (`std::bad_alloc`) into some deadlock.

denesb commented:
https://github.com/scylladb/scylla/pull/5343/commits/8c5ee56c6825fbdd8c1022fcd74d283679755dfe should be folded into the commit it fixes.

denesb commented on db/view/view_update_generator.cc:
```diff
@@ -39,10 +39,25 @@ future<> view_update_generator::start() {
+                auto& entry = it->second;
+                try {
+                    entry.t->move_sstables_from_subdirs(entry.sstables).get();
+                } catch (...) { /* ignore */ }
```
Why are we ignoring this?

bhalevy replied:
1. If moving sstables from the staging directory fails, they will be retried on restart.
2. Individual errors are already logged by move_sstables_from_subdirs so no need for another log message here.

I'll add that to a comment...

denesb commented on sstables/sstables.cc:
```diff
@@ -2679,20 +2679,24 @@ future<> sstable::set_generation(int64_t new_generation) {
+            return remove_file(sstable::filename(old_dir, _schema->ks_name(), _schema->cf_name(), _version, old_generation, _format, p.second));
+        });
+    }).then([this, old_dir, new_dir] {
+        return when_all_succeed(sync_directory(old_dir), sync_directory(new_dir));
```
Can this `sync_directory()` throw? If the first one that is called throws, the second one won't even be invoked.

denesb replied:
Thinking about it, this was also the case before, so I guess its fine.

bhalevy replied:
Correct

denesb commented on table.cc:
```diff
@@ -2476,15 +2476,29 @@ table::make_reader_excluding_sstable(schema_ptr s,
+                        return make_ready_future<>();
+                    } else {
+                        auto ep = f.get_exception();
+                        tlogger.error("Failed to move sstable {} from staging: {}", sst->get_filename(), ep);
```
The message has `staging` hardcoded as the subdir, while we can have other subdirs here, no?

bhalevy replied:
True. Will fix.

denesb commented on table.cc:
```diff
@@ -2476,15 +2476,29 @@ table::make_reader_excluding_sstable(schema_ptr s,
+                    if (!f.failed()) {
+                        _sstables_staging.erase(sst->generation());
+                        _compaction_strategy.get_backlog_tracker().add_sstable(sst);
+                        dirs_to_sync.emplace(sst->get_dir());
```
Won't this always be the sstable's new directory (after the move)?

bhalevy replied:
Yes. Good catch.

denesb commented on table.cc:
```diff
@@ -2476,15 +2476,29 @@ table::make_reader_excluding_sstable(schema_ptr s,
+            return parallel_for_each(std::move(sstables), [this, &dirs_to_sync] (sstables::shared_sstable sst) {
+                return sst->move_to_new_dir(dir(), sst->generation(), false).then_wrapped([this, sst, &dirs_to_sync] (future<> f) {
+                    if (!f.failed()) {
+                        _sstables_staging.erase(sst->generation());
```
This assumes this sstable is moved from the staging dir. Aren't there other subdirs we could be moving from?

bhalevy replied:
Hmm, not at this time but you're right.
This function was previously called `move_sstable_from_staging_in_thread` and I made it general.
I'll fix that.

bhalevy commented:
In v3:
- handle exception in `view_update_generator::start` and log appropriate messages.
- fix `dirs_to_sync.emplace(sst->get_dir())` in `table::move_sstables_from_staging`
  - move before `sst->move_to_new_dir`
- use name binding for `sstables_with_tables` fields and for `sstables_to_move` iterator pair.
- `move_sstables_from_staging` is specific to staging
  rather than calling it `move_sstables_from_subdirs`

denesb commented on db/view/view_update_generator.cc:
```diff
@@ -33,13 +35,23 @@ future<> view_update_generator::start() {
+              } catch(...) {
+                // Move from staging will be retried upon restart.
+                vug_logger.warn("Moving {} from staging failed: {}. Ignoring...", entry.sst->get_filename(), std::current_exception());
+              }
```
Now we will execute `pop_front()` and `signal()` below even if processing this sstable failed, which means this will be retried on restart.

I wonder if it would be better to just leave the entry and the semaphore, and just do a `continue` in the case of an error, so that the loop retries on the next iteration. In this case it would be best to use `semaphore_units` and add it to the `_sstables_with_tables`, letting `pop_front()` automatically signal the semaphore, instead of doing it manually.

BTW by "not in this patch" below I meant not in this PR. The rest of this PR looks good, I think this unrelated fix could be fixed in a separate one, allowing this PR to go in.

bhalevy replied:
@denesb, thanks for your review.

> Now we will execute pop_front() and signal() below even if processing this sstable failed, which means this will be retried on restart.

Actually, if processing stage throws, we break out of the loop so the entry will remain in `_sstables_with_tables` and will be retried on the next iteration.
Only moving out of the staging sub-directory is ignored and will be retried on restart.

denesb replied:
Ah, I see.

bhalevy commented:
ping @nyh

slivne commented:
@nyh ping

psarna commented on db/view/view_update_generator.hh:
```diff
@@ -42,9 +42,10 @@ class view_update_generator {
+        sstable_with_table(sstables::shared_sstable sst, lw_shared_ptr<table> t) : sst(std::move(sst)), t(std::move(t)) { }
     };
     std::deque<sstable_with_table> _sstables_with_tables;
+    std::unordered_map<lw_shared_ptr<table>, std::vector<sstables::shared_sstable>> _sstables_to_move;
```
Is there a reason this one is a map, instead of simply being a deque, just like `_sstables_with_tables`? It seems like it's only ever used to iterate over all of its elements, and deque looks like a more natural structure for that access pattern (no per-item allocations, less overhead, and so on).

bhalevy replied:
@psarna it is a map since it is 2-dimensional. I keep a vector of sstables per table (aka column_family) and then call t->move_sstables_from_staging on that vector.  The reason is that they are all stored in the same staging directory so we can sync it and its parent at the end after moving all sstables out of it, rather than syncing at random for each individual sstable.

psarna commented on table.cc:
```diff
@@ -2476,15 +2476,29 @@ table::make_reader_excluding_sstable(schema_ptr s,
-    _compaction_strategy.get_backlog_tracker().add_sstable(sst);
+future<> table::move_sstables_from_staging(std::vector<sstables::shared_sstable> sstables) {
+    return with_semaphore(_sstable_deletion_sem, 1, [this, sstables = std::move(sstables)] {
+        return do_with(std::set<sstring>({dir()}), [this, sstables = std::move(sstables)] (auto& dirs_to_sync) {
```
nitpick: `auto&` parameters like this tend to confuse my IDE, could you replace it with `std::set<sstring>&`?

psarna commented on sstables/sstables.cc:
```diff
@@ -2679,20 +2679,24 @@ future<> sstable::set_generation(int64_t new_generation) {
+            }
+            return remove_file(sstable::filename(old_dir, _schema->ks_name(), _schema->cf_name(), _version, old_generation, _format, p.second));
+        });
+    }).then([this, old_dir, new_dir] {
```
[no action needed] by now you could std::move the string, but perhaps it's better to leave it as is, in case somebody adds a continuation later.

psarna commented on table.cc:
```diff
@@ -2476,15 +2476,29 @@ table::make_reader_excluding_sstable(schema_ptr s,
+future<> table::move_sstables_from_staging(std::vector<sstables::shared_sstable> sstables) {
+    return with_semaphore(_sstable_deletion_sem, 1, [this, sstables = std::move(sstables)] {
+        return do_with(std::set<sstring>({dir()}), [this, sstables = std::move(sstables)] (auto& dirs_to_sync) {
+            return parallel_for_each(std::move(sstables), [this, &dirs_to_sync] (sstables::shared_sstable sst) {
```
Have you considered how high can parallelism get here? My question is: can there be, say, 100k staging tables queued for moving from staging at a time, e.g. because something keeps failing and the queue grows and grows? Because if that's plausible, then we have a problem, especially that `move_to_new_dir` also leverages `parallel_for_each` in multiple places, and we could end up with millions of continuations getting created here each time this function runs.

I hope the scenario above is impossible and I just missed some details, but I'd like a confirmation :)

bhalevy replied:
It's unlikely, but with a configuration of thousands of tables having views we may end up with thousands of sstables to move.

bhalevy replied:
Since we're now syncing the directories at the end we can just move the sstables one at a time. I don't think that parallelism will buy us much in this case.

avikivity replied:
It does get us something. If the server is busy each syscall operation ends up taking half a millisecond, and a move may be several such non-parallel operations. So moving tens of thousands of sstables can take tens of seconds.

Ok, maybe this isn't so important.

bhalevy replied:
also. please note we do the move operations on all the components of each sstable in parallel in `sstable::create_links` and then in `sstable::move_to_new_dir`

bhalevy commented on db/view/view_update_generator.cc:
```diff
@@ -32,17 +34,39 @@ future<> view_update_generator::start() {
+                try {
+                    // collect all staging sstables to move in a map, grouped by table.
+                    _sstables_to_move[t].push_back(sst);
+                } catch(...) {
```
sure.

bhalevy commented:
In v2:
- nit: added space `s/catch(/catch (/`
- nit: fixed typo in commit `s/nder/under/`
- in do_with: `s/auto&/std::set<sstring>&/`
- in move_sstables_from_staging: `s/parallel_for_each/do_for_each/` so we move sstables sequentially rather than in parallel assuming write behind on directories and single flush at the end.

nyh commented:
@bhalevy, @psarna is on vacation. The code looks good to me, and it appears you addressed @psarna's concerns - I'll commit your code now and when @psarna returns if he wants more changes he can ask for them then.
